### ISSUE TITLE
Improvements

### ISSUE THREAD

**Issue by renard** [skx/implant#9]
Hi,

I found `implant` after looking for `esc` alternatives since it is overkill for my needs. I also dieted final binaries by 4Mb.

After a quick glance here is what I found that could be useful. Just hopping to not break anything.

1. Use `gzip.BestCompression`: since the static file is generated once, there is no real overhead for that. The result is smaller binaries.

2. Instead of storing files in an array, it would be better to use a map since the lookup would be faster.

3. `append` is a costly operation since the whole array has to be recreated. The number of files is known it built time. Maybe creating a static array in the template would be more efficient.

This is just a few thoughts that I have and haven't yet tried to do anything in that way. Let me know if you are interested in.

Cheers.

**Comment by skx** [comment]
They seem like good suggestions for change.   If you implemented some/all of them in (separate) pull-requests, which did not break the existing tests I would be pleased and would accept your contribution.

If not I'll take a look at them over the next week or two.

**Comment by renard** [comment]
Ok nice.

BTW I add a simple PR to use BestCompression instead of BestSpeed and the tests failed.
I don't really understand why.

OTOH in `finder/finder_test.go` this check fails on MacOS since `/etc/fstab` does not exist:
```golang
        if !finder.ShouldInclude("/etc/fstab") {
                t.Fatalf("We should include normal files!")
        }
```

**Comment by skx** [comment]
Great, thanks for the PR.  Once I've eaten I'll take a look - and I'll fix the test(s) before tomorrow morning.

**Comment by renard** [comment]
Yes maybe

```golang
if !finder.ShouldInclude("/bin/sh") {
```

is a better alternative. I can provide the PR if you like.

**Comment by skx** [comment]
I picked `/etc/passwd`.  Ideally I'd ship a top-level `_tests` directory in the repo and use that.  I'll switch to doing so shortly, but the `passwd` change is a quick fix for the moment.

**Comment by renard** [comment]
Not sure this is a good idea to use `/etc/passwd`  for testing purposes since it may contain sensible data.

**Comment by skx** [comment]
Will switch in the next hour or two.  Food.

**Comment by renard** [comment]
I have pushed #14 this should do the job.

BTW I faced a chicken-and-egg problem since `static.go` which contains `data/static.tmpl` is generated from ... wait for it ...`data/static.tmpl`.

One tedious solution is to run:
```shell
cat data/static.tmpl | gzip -9 | xxd -p | tr -d '\n'
```

I am thinking about storing the last modification time and add a functionality to use local file if found and newer than the compiled one. What do you think?

**Comment by skx** [comment]
> I am thinking about storing the last modification time and add a functionality to use local file if found and newer than the compiled one. What do you think?

This I wouldn't do.  Yes it's a pain when the formatting/implementation changes.  But Perl can fix that ;)

Really this is one of the things I like about the tool, it is simple.  There are some similar projects which allow handling local-files or templates, etc, and I think that's not what this project does.  The same reason I don't implement a HTTP-server interface, or fs-interface.  Just access to resource by name.

**Comment by renard** [comment]
Ok fair enough. 
Still having a timestamp would be a hint for higher level usage.
If you don't feel confortable with this option you can close this PR since it already solves the issue I was facing to :-)

**Comment by skx** [comment]
I will close that one then, thanks.

I appreciate your contributions and I guess this issue can be closed too now.

I'll make a new release before the weekend.

**Comment by renard** [comment]
You're welcome ;-)

I introduced a nasty bug (which is fixed in #17 ) with an unclosed stream.

FYI on some other project Hex vs. Base64 encoding + gzip.BestCompression reduced the size by  at least 30% :-)

**Comment by skx** [comment]
Closing this now the contributions have been included, and a new release made.

* https://github.com/skx/implant/releases/tag/release-0.6

Thanks again!  Any further ideas/suggestions/contributions would be welcome :)
